File: src/types/schemas.ts
/**
 * NutriVision AI — TypeScript schema contracts
 *
 * Mirrors the planned `app/models/schemas.py` (FastAPI/Pydantic) on the
 * backend. This file is the **source of truth** for the frontend.
 *
 * Per the blueprint's Section 8 directive, any backend schema change
 * requires this file to update in lockstep to avoid integration drift.
 *
 * Naming convention:
 *   - PascalCase types
 *   - snake_case fields (matches what a Python backend would emit)
 *   - All numeric nutrient values are per-100g servings unless suffixed
 */

// ─────────────────────────────────────────────────────────────────────────
//  Core nutrition primitives
// ─────────────────────────────────────────────────────────────────────────

/** Per-100g macro/micro nutrient facts for a single food item. */
export interface NutritionFacts {
  /** grams of protein */
  protein: number;
  /** grams of carbohydrates */
  carbs: number;
  /** grams of fat */
  fat: number;
  /** kilocalories */
  kcal: number;
  /** grams of fiber */
  fiber: number;
  /** milligrams of sodium */
  sodium: number;
  /** grams of sugar */
  sugar: number;
  /** glycemic share, 0–1 (used by personalize filter) */
  glycemic: number;
}

/** Food category used to pick tile gradients and grocery group routing. */
export type FoodCategory = 'protein' | 'produce' | 'grains' | 'other';

/** A single food with display metadata + per-100g nutrition. */
export interface FoodDescriptor extends NutritionFacts {
  emoji: string;
  category: FoodCategory;
  /** Alternative names the vision model may also emit. */
  alternatives: string[];
  /** Pairings the grocery list may suggest. */
  pairings: string[];
}

// ─────────────────────────────────────────────────────────────────────────
//  Vision pipeline outputs
// ─────────────────────────────────────────────────────────────────────────

/**
 * One food item detected by the dual vision pipeline (Gemini + HF validator)
 * before nutrition is resolved.
 */
export interface DetectedItem {
  /** Canonical food name; should match a key in NUTRITION_DB */
  name: string;
  /** Confidence 0–100; <70 is flagged low-confidence */
  confidence: number;
  /** Estimated portion size in grams */
  grams: number;
  /** Optional user-facing note (e.g. "Low confidence — tap to confirm") */
  note?: string | null;
  /** Optional bounding box from the vision model [x, y, w, h] in 0–1 space */
  bbox?: [number, number, number, number];
}

/**
 * A DetectedItem after passing through the nutrition DB cascade (OFF →
 * USDA → Fruityvice). Used by personalize to re-rank flagged items.
 */
export interface ResolvedItem extends DetectedItem {
  /** Resolved per-100g facts; null if no DB had a confident match */
  nutrition: NutritionFacts | null;
  /** Display metadata resolved along with nutrition */
  descriptor: FoodDescriptor | null;
  /** Which DB resolved this item (for explainability) */
  source: 'open-food-facts' | 'usda' | 'fruityvice' | 'cache' | 'estimated';
  /** True if the resolver only matched an "estimated" entry */
  partial?: boolean;
}

/**
 * Per-nutrient, per-item contribution. Powers the SHAP-style explainability
 * panel on NutrientsScreen.
 */
export interface NutrientContribution {
  /** Which nutrient is being explained */
  nutrient: 'sodium' | 'sugar' | 'fiber' | 'protein' | 'carbs' | 'fat' | 'kcal';
  /** Food item contributing */
  itemName: string;
  /** Absolute amount in the same unit as NutritionFacts */
  amount: number;
  /** Share of total in this nutrient (0–1) */
  share: number;
  /** True if this contribution crosses a flag threshold for the user */
  flagged: boolean;
  /** Human-readable reason this was flagged (item-specific, used by verdict chips) */
  reason?: string;
  /** Visual tone for the verdict chip — 'warn' for over-target, 'good' for under-target supportive */
  tone?: 'good' | 'warn';
}

/**
 * A pipeline trace entry surfaced in a tiny toast during analysis. Proves
 * the multi-stage architecture is real, not just a stub.
 */
export interface PipelineTrace {
  stage: 'capture' | 'cache-check' | 'vision-id' | 'hf-validate' | 'reconcile' | 'nutrition' | 'portion' | 'personalize' | 'explain';
  status: 'pending' | 'running' | 'done' | 'cache-hit' | 'fallback';
  startedAt: number;
  finishedAt?: number;
  label: string;
}

export interface AnalyzeResult {
  detected: DetectedItem[];
  pipeline: PipelineTrace[];
}

export interface ResolveNutritionResult {
  resolved: ResolvedItem[];
  contributions: NutrientContribution[];
}

// ─────────────────────────────────────────────────────────────────────────
//  User profile / personalization
// ─────────────────────────────────────────────────────────────────────────

/** Coarse metabolic goal the user can toggle. */
export interface HealthGoals {
  diabetic: boolean;
  protein: boolean;
  budget: boolean;
  mediter: boolean;
}

/** Dietary preferences (vegetarian, vegan, etc.). */
export type DietaryPreference =
  | 'Vegetarian'
  | 'Vegan'
  | 'Pescatarian'
  | 'Gluten-free'
  | 'Dairy-free'
  | 'Halal'
  | 'Kosher';

/** Common allergens tracked in the profile. */
export type Allergen =
  | 'Peanuts'
  | 'Tree nuts'
  | 'Shell shellfish'
  | 'Soy'
  | 'Eggs'
  | 'Dairy'
  | 'Gluten'
  | 'Sesame';

export interface UserProfile {
  name: string;
  goals: HealthGoals;
  preferences: DietaryPreference[];
  allergens: Allergen[];
  /** Daily grocery budget ceiling, in local currency (BDT).
   *  The Grocery screen applies a 7× multiplier to get the weekly cap. */
  budget: number;
  /** Number of people the meal serves */
  serving: number;
  /** Daily kcal target for the Home progress bar. Optional for backward
   *  compatibility — older profiles stored before this shipped will fall
   *  back to `DAILY_KCAL_TARGET_DEFAULT` (2000) at read time. */
  dailyKcalTarget?: number;
}

/** Universal fallback daily kcal target. Roughly the adult average used by
 *  US dietary guidelines (2000 kcal/day). Profiles that opt-in can override
 *  via the Profile screen; until then we use this so the Home progress bar
 *  always renders a meaningful percentage. */
export const DAILY_KCAL_TARGET_DEFAULT = 2000;

/** Resolve a profile's effective daily kcal target. Tolerates older stored
 *  profiles that pre-date the `dailyKcalTarget` field by falling back to the
 *  universal default. Always returns a positive finite number. */
export function resolveDailyKcalTarget(profile: UserProfile | null | undefined): number {
  const v = profile?.dailyKcalTarget;
  if (typeof v === 'number' && Number.isFinite(v) && v > 0) return v;
  return DAILY_KCAL_TARGET_DEFAULT;
}

/** Per-meal nutrition snapshot — mirrors backend `MealTotals`. */
export interface MealTotals {
  kcal: number;
  protein: number;
  carbs: number;
  fat: number;
  fiber: number;
  /** milligrams of sodium */
  sodium: number;
  sugar: number;
}

/**
 * Today's logged-meals summary — the shape the Home progress bar consumes.
 * Mirrors backend `TodaySummary`.
 */
export interface TodaySummary {
  mealsLogged: number;
  totalKcal: number;
  totalProtein: number;
  totalCarbs: number;
  totalFat: number;
  dailyKcalTarget: number;
}

// ─────────────────────────────────────────────────────────────────────────
//  Grocery list
// ─────────────────────────────────────────────────────────────────────────

export interface GroceryItem {
  id: string;
  name: string;
  /** Per-unit price in BDT */
  price: number;
  checked: boolean;
}

export interface GroceryGroup {
  category: string;
  items: GroceryItem[];
}

export interface GroceryList {
  budget: number;
  groups: GroceryGroup[];
}

// ─────────────────────────────────────────────────────────────────────────
//  Tabs (client-side)
// ─────────────────────────────────────────────────────────────────────────

export type TabId = 'capture' | 'results' | 'nutrients' | 'profile' | 'grocery';

export interface TabDescriptor {
  id: TabId;
  label: string;
  icon: string;
}

// ─────────────────────────────────────────────────────────────────────────
//  Auth (mock-grade for the demo)
// ─────────────────────────────────────────────────────────────────────────

/** Role determines which app surface the session lands on. The current
 *  hackathon build only ships `user` (regular PWA) and `admin` (dashboard). */
export type Role = 'user' | 'admin';

/** Lightweight session record persisted to localStorage under `nv_auth`.
 *  Passwords are not stored — the auth lib matches by username prefix and
 *  password is only used to gate a non-empty form submission. */
export interface AuthSession {
  email: string;
  role: Role;
  /** ISO timestamp when the session was created. */
  loginAt: string;
}

/** Synthetic account the admin dashboard sees in its directory. Admins do
 *  not see user passwords — only the public profile + a generated meal
 *  history. The real signed-in user is also surfaced as one of these. */
export interface DemoUser {
  email: string;
  displayName: string;
  /** Days since "joining" the platform — informs the joined-on chip. */
  joinedDaysAgo: number;
  meals: MealEntry[];
}

// ─────────────────────────────────────────────────────────────────────────
//  Helpers
// ─────────────────────────────────────────────────────────────────────────

/** Rough confidence bucket used for badge styling on ResultsScreen. */
export type ConfidenceBucket = 'high' | 'medium' | 'low';

export function bucketizeConfidence(c: number): ConfidenceBucket {
  if (c >= 80) return 'high';
  if (c >= 60) return 'medium';
  return 'low';
}

// ─────────────────────────────────────────────────────────────────────────
//  Meal history
// ─────────────────────────────────────────────────────────────────────────

/** A single logged meal stored in localStorage history. */
export interface MealEntry {
  /** Unique ID (timestamp-based) */
  id: string;
  /** ISO 8601 timestamp */
  date: string;
  /** Meal label (auto-generated from items) */
  label: string;
  /** Detected food items at the time of logging */
  items: DetectedItem[];
  /** Snapshot of nutrition totals (kcal, protein, carbs, fat, fiber, sodium, sugar) */
  totals: MealTotals;
  /** Optional captured photo data URL */
  photoUrl?: string | null;
  /** Which profile goals were active when this meal was logged */
  activeGoals: HealthGoals;
}

/** Aggregated stats for the profile screen. */
export interface UserStats {
  totalMeals: number;
  totalDaysActive: number;
  avgDailyKcal: number;
  streakDays: number;
}

File: app/models/schemas.py
"""Pydantic v2 schemas — 1:1 mirror of `src/types/schemas.ts`.

Lockstep rule (from `NutriVision_AI_Blueprint.md` §8):
    Any backend schema change requires the matching TS interface in
    `src/types/schemas.ts` to update in the same commit, and vice
    versa. Treat schema drift as a breaking change.

Field naming: the frontend already uses snake_case (`daily_kcal_target`,
`active_goals`, `meal_logged`), so this module is a direct port. No
alias gymnastics needed for the common case.

Where the TS surface uses inline shapes that aren't named (e.g.
`getTodaySummary()`'s return type), we lift them into named models
here (`TodaySummary`) so the FastAPI response_model is explicit.
"""

from __future__ import annotations

from typing import Literal

from pydantic import BaseModel, ConfigDict, Field, field_validator


# ─────────────────────────────────────────────────────────────────────────
# Core nutrition primitives
# ─────────────────────────────────────────────────────────────────────────


class NutritionFacts(BaseModel):
    """Per-100g macro/micro nutrient facts for a single food item."""

    protein: float
    carbs: float
    fat: float
    kcal: float
    fiber: float
    sodium: float
    sugar: float
    """Glycemic share, 0–1 (used by personalize filter)."""
    glycemic: float


FoodCategory = Literal["protein", "produce", "grains", "other"]


class FoodDescriptor(NutritionFacts):
    """A single food with display metadata + per-100g nutrition."""

    emoji: str
    category: FoodCategory
    """Alternative names the vision model may also emit."""
    alternatives: list[str]
    """Pairings the grocery list may suggest."""
    pairings: list[str]


# ─────────────────────────────────────────────────────────────────────────
# Vision pipeline outputs
# ─────────────────────────────────────────────────────────────────────────


# `bbox` is a fixed-length 4-tuple of normalized [x, y, w, h] floats.
BBox = tuple[float, float, float, float]


class DetectedItem(BaseModel):
    """One food item detected by the dual vision pipeline (Gemini + HF
    validator) before nutrition is resolved."""

    name: str
    """Confidence 0–100; <70 is flagged low-confidence."""
    confidence: float
    """Estimated portion size in grams."""
    grams: float
    """Optional user-facing note (e.g. 'Low confidence — tap to confirm')."""
    note: str | None = None
    """Optional bounding box from the vision model [x, y, w, h] in 0–1
    normalized coordinates."""
    bbox: BBox | None = None

    @field_validator("bbox")
    @classmethod
    def _bbox_length(cls, v: BBox | None) -> BBox | None:
        """Reject malformed bounding boxes (anything other than 4 floats)."""
        if v is not None and len(v) != 4:
            raise ValueError("bbox must be exactly 4 floats: (x, y, w, h)")
        return v


class ResolvedItem(DetectedItem):
    """A `DetectedItem` after passing through the nutrition DB cascade."""

    nutrition: NutritionFacts | None = None
    descriptor: FoodDescriptor | None = None
    """Which DB resolved this item (for explainability)."""
    source: Literal[
        "open-food-facts", "usda", "fruityvice", "cache", "estimated"
    ]
    """True if the resolver only matched an 'estimated' entry."""
    partial: bool = False


# Pipeline stage + status mirrors the TS `PipelineTrace['stage'|'status']`
# literals exactly — these are also used by the frontend CaptureScreen
# toast to label each stage.
PipelineStage = Literal[
    "capture",
    "cache-check",
    "vision-id",
    "hf-validate",
    "reconcile",
    "nutrition",
    "portion",
    "personalize",
    "explain",
]
PipelineStatus = Literal[
    "pending", "running", "done", "cache-hit", "fallback"
]


class PipelineTrace(BaseModel):
    """A pipeline trace entry surfaced in a tiny toast during analysis.
    Proves the multi-stage architecture is real, not just a stub."""

    stage: PipelineStage
    status: PipelineStatus
    started_at: int
    """Wall-clock ms; matches `finishedAt` field in the TS interface."""
    finished_at: int | None = None
    label: str

    # Allow the snake_case ↔ camelCase aliases so older payloads still
    # parse while we migrate call sites. Both forms are accepted.
    model_config = ConfigDict(populate_by_name=True)


class AnalyzeResult(BaseModel):
    detected: list[DetectedItem]
    pipeline: list[PipelineTrace]


class NutrientContribution(BaseModel):
    nutrient: Literal[
        "sodium", "sugar", "fiber", "protein", "carbs", "fat", "kcal"
    ]
    item_name: str
    amount: float
    """Share of total in this nutrient (0–1)."""
    share: float
    flagged: bool
    reason: str | None = None
    """Visual tone for the verdict chip."""
    tone: Literal["good", "warn"] | None = None


class ResolveNutritionResult(BaseModel):
    resolved: list[ResolvedItem]
    contributions: list[NutrientContribution]


# ─────────────────────────────────────────────────────────────────────────
# User profile / personalization
# ─────────────────────────────────────────────────────────────────────────


class HealthGoals(BaseModel):
    diabetic: bool = False
    protein: bool = False
    budget: bool = False
    mediter: bool = False


DietaryPreference = Literal[
    "Vegetarian",
    "Vegan",
    "Pescatarian",
    "Gluten-free",
    "Dairy-free",
    "Halal",
    "Kosher",
]

Allergen = Literal[
    "Peanuts",
    "Tree nuts",
    "Shell shellfish",
    "Soy",
    "Eggs",
    "Dairy",
    "Gluten",
    "Sesame",
]


class UserProfile(BaseModel):
    name: str
    goals: HealthGoals
    preferences: list[DietaryPreference]
    allergens: list[Allergen]
    """Daily grocery budget ceiling, in local currency (BDT)."""
    budget: float
    """Number of people the meal serves."""
    serving: int
    """Daily kcal target for the Home progress bar. Optional — older
    profiles that pre-date this field will fall back to the universal
    default (`DAILY_KCAL_TARGET_DEFAULT`)."""
    daily_kcal_target: float | None = Field(default=None, alias="dailyKcalTarget")

    model_config = ConfigDict(populate_by_name=True)


# Universal fallback daily kcal target (mirrors the TS module constant).
DAILY_KCAL_TARGET_DEFAULT: float = 2000


# ─────────────────────────────────────────────────────────────────────────
# Grocery list
# ─────────────────────────────────────────────────────────────────────────


class GroceryItem(BaseModel):
    id: str
    name: str
    """Per-unit price in BDT."""
    price: float
    checked: bool = False


class GroceryGroup(BaseModel):
    category: str
    items: list[GroceryItem]


class GroceryList(BaseModel):
    budget: float
    groups: list[GroceryGroup]


# ─────────────────────────────────────────────────────────────────────────
# Meal history + aggregates
# ─────────────────────────────────────────────────────────────────────────


class MealTotals(BaseModel):
    """Snapshot of nutrition totals (kcal, protein, carbs, fat, fiber,
    sodium, sugar) for a logged meal."""

    kcal: float
    protein: float
    carbs: float
    fat: float
    fiber: float
    sodium: float
    sugar: float


class MealEntry(BaseModel):
    """A single logged meal stored in history."""

    id: str
    """ISO 8601 timestamp."""
    date: str
    """Auto-generated meal label."""
    label: str
    items: list[DetectedItem]
    totals: MealTotals
    photo_url: str | None = Field(default=None, alias="photoUrl")
    """Which profile goals were active when this meal was logged."""
    active_goals: HealthGoals

    model_config = ConfigDict(populate_by_name=True)


class UserStats(BaseModel):
    total_meals: int
    total_days_active: int
    avg_daily_kcal: int
    streak_days: int


# Lifted from `getTodaySummary()`'s inline return type in the frontend
# so FastAPI has an explicit response_model.
class TodaySummary(BaseModel):
    meals_logged: int
    total_kcal: float
    total_protein: float
    total_carbs: float
    total_fat: float
    daily_kcal_target: float


# ─────────────────────────────────────────────────────────────────────────
# Targets — daily nutrient ceilings used by the personalize filter.
# Matches `MOCK_TARGETS` in `src/mocks/fixtures.ts`.
# ─────────────────────────────────────────────────────────────────────────


class MacroTargets(BaseModel):
    sodium: float
    sugar: float
    fiber: float
    protein: float
    carbs: float
    fat: float
    kcal: float
